### PR TITLE
[16.0][MIG] delivery_carrier_package_measure_required: Migration to 16.0

### PR DESCRIPTION
**Pull request by RodrigoBM**
issue #534 

Migration from 14.0

### CODE REVIEW

File: delivery_carrier_package_measure_required/tests/test_delivery_carrier_package_measure_require.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            {
+                "product_id": cls.product_aw.id,
+                "product_uom_id": cls.uom_kg.id,
+                "picking_id": cls.pick.id,
+                "qty_done": 5,
+                "location_id": cls.stock_location.id,
+                "location_dest_id": cls.customer_location.id,
+            }
+        )
+        cls.pick.action_confirm()
```
Comment (xAdrianCif) [inline delivery_carrier_package_measure_required/tests/test_delivery_carrier_package_measure_require.py]: I know that this is a migration and comes from there, but to respect the stock flow in Odoo, it's better to create a `stock.move` (instead of a `stock.move.line`) directly associated with `stock.picking`, and then force availability using the 'action_assign' method. If there is availability, the `stock.move.line` is created, and we can modify its `qty_done`.

Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/tests/test_delivery_carrier_package_measure_require.py]: done

**Review by xAdrianCif — APPROVED** [review]
Tech review, thanks for the job!

**Comment by RodrigoBM** [comment]
is it possible to merge this, @OCA/logistics-maintainers, thank you

**Comment by RodrigoBM** [comment]
ping @TDu  @rousseldenis @simahawk @jbaudoux @lmignon

**Review by jorgemartinez-factorlibre — APPROVED** [review]
Tech review, good job @RodrigoBM

**Comment by hugosantosred** [comment]
@OCA/logistics-maintainers Could you review for merge or apply the no stale label to this PR? Thanks!

**Comment by simahawk** [comment]
@RodrigoBM can you rebase pls?

**Review by simahawk — COMMENTED** [review]
@TDu ping for review ;)

**Comment by RodrigoBM** [comment]
> @RodrigoBM can you rebase pls?

Done @simahawk

File: delivery_carrier_package_measure_required/models/stock_picking.py [hunk 2]
```diff
@@ -19,3 +19,23 @@ def _check_required_package_measurement(self):
...
+        res.write(
+            {
+                "pack_length": package_length,
+                "width": package_width,
+                "height": package_height,
+                "pack_weight": package_weight,
+            }
+        )
+
+        return res
```
Comment (jbaudoux) [inline delivery_carrier_package_measure_required/models/stock_picking.py:37]: I guess not part of the migration but an improvement. If yes, can it go in a separate commit?

File: delivery_carrier_package_measure_required/models/stock_quant_package.py [hunk 3]
```diff
@@ -8,20 +8,23 @@
...
         ignore_pack_content = self.env.context.get(
-            "delivery_pkg_measure__ignore_package_content"
+            "delivery_pkg_measure__ignore_package_content", False
+        )
+        force_validation = self.env.context.get(
+            "delivery_pkg_measure__force_validation_package", False
         )
         for package in self:
-            if not ignore_pack_content and not package.quant_ids:
+            if ignore_pack_content or (not force_validation and not package.quant_ids):
```
Comment (jbaudoux) [inline delivery_carrier_package_measure_required/models/stock_quant_package.py]: I think we better stick to previously existing context variable (and not invert behavior). I don't find it clearer with the new name.

File: delivery_carrier_package_measure_required/models/stock_picking.py (hunk 2 above)
Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_picking.py:37]: Done.

File: delivery_carrier_package_measure_required/models/stock_quant_package.py (hunk 3 above)
Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_quant_package.py]: @jbaudoux , I think this is better now.

Before the changes we don't check the dimensions in the case if ignore_pack_content is False and the package has no quants.
We also have a variable that is passed by context which is delivery_pkg_measure__ignore_package_content that if you pass it with the value set to True you would expect it not to check the dimensions but this is not the case.

Option1:
ignore_pack_content = True, package.quant_ids = True
(not ignore_pack_content and not package.quant_ids) if == False --> Check dimensions
Opción 2:
ignore_pack_content = True, package.quant_ids = False
(not ignore_pack_content and not package.quant_ids) if == False --> Check dimensions
Opción 3:
ignore_pack_content = False, package.quant_ids = True
(not ignore_pack_content and not package.quant_ids) if == False --> Check dimensions
Opción 4:
ignore_pack_content = False, package.quant_ids = False
(not ignore_pack_content and not package.quant_ids) if == True --> No Check dimensions

But what if we want to be able to skip the dimension check in other cases?

Option1:
ignore_pack_content = False, force_validation = True, package.quant_ids = True
ignore_pack_content or (not force_validation and not package.quant_ids) if = False --> Check dimensions
Opción 2:
ignore_pack_content = False, force_validation = True, package.quant_ids = False
ignore_pack_content or (not force_validation and not package.quant_ids) if = False --> Check dimensions
Opción 3:
ignore_pack_content = False, force_validation = False, package.quant_ids = True
ignore_pack_content or (not force_validation and not package.quant_ids) if = False --> Check dimensions
Opción 4:
ignore_pack_content = False, force_validation = False, package.quant_ids = False
ignore_pack_content or (not force_validation and not package.quant_ids) if = True --> No Check dimensions
Opción 5:
ignore_pack_content = True
ignore_pack_content or (not force_validation and not package.quant_ids) if = True --> No Check dimensions

What do you think about this? this can go in another commit if you think it is more correct.

Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_quant_package.py]: however, the delivery_pkg_measure__ignore_package_content context I have not seen used in other OCA modules.

Comment (jbaudoux) [reply delivery_carrier_package_measure_required/models/stock_quant_package.py]: I can't find any usage of this context variable. @TDu What was the purpose of it?
Do we need any of those context variables? If this is useless, I'm more in favor of simplifying. We can already configure which dimension is required.

Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_quant_package.py]: I understand @jbaudoux , but now I have client developments that need these contexts. We can change this "if ignore_pack_content or (not force_validation and not package.quant_ids):" to "if self.ignore_check_dimensions():" and have everyone inherit the method to whatever we need.

Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_quant_package.py]: ping @TDu

Comment (TDu) [reply delivery_carrier_package_measure_required/models/stock_quant_package.py]: As commented in the code that content key `delivery_pkg_measure__ignore_package_content`  was meant to be used to force asking for dimension even for empty packages.

I think the idea behind it was, you receive a new product in your warehouse, all packaging (A, B) for that product are not configured properly. The product arrives in packaging A. But you want the user to configure all packaging (A and B) for that product on first reception.

Now indeed it does not look like it is used :thinking:

File: delivery_carrier_package_measure_required/models/stock_picking.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _check_required_package_measurement(self):
+        """Check the required measurement on destination pacakges.
+
+        If package is None all result packages on the stock pickings are checked
+        otherwise only the one
```
Comment (TDu) [inline delivery_carrier_package_measure_required/models/stock_picking.py]:
```suggestion
```
docstring out of sync with the code ?

Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_picking.py]: Now that I think about it, I think so

Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_picking.py]: Done

File: delivery_carrier_package_measure_required/models/stock_quant_package.py (hunk 3 above)
Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_quant_package.py]: Is it a problem to accept this new proposal? Do you need it to be in a separate commit?

File: delivery_carrier_package_measure_required/models/stock_picking.py (hunk 4 above)
Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_picking.py]: this is now resolved, @TDu?

File: delivery_carrier_package_measure_required/models/stock_picking.py (hunk 2 above)
Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_picking.py:37]: this is now resolved, @jbaudoux?

**Comment by RodrigoBM** [comment]
ping @TDu @jbaudoux

File: delivery_carrier_package_measure_required/models/stock_quant_package.py (hunk 3 above)
Comment (jbaudoux) [reply delivery_carrier_package_measure_required/models/stock_quant_package.py]: Thanks for the clarification @TDu 
@RodrigoBM Your current proposal is fine for me. It is clear and covers all the cases.
Just put that in a separate commit :)

Comment (RodrigoBM) [reply delivery_carrier_package_measure_required/models/stock_quant_package.py]: Done @jbaudoux, sorry for the delay

**Comment by jbaudoux** [comment]
@RodrigoBM As a general rule, can you keep commit subject (first line) short and put description in the body (separated by a blank line from the subject).

Like:
```
delivery_carrier_package_measure_required: put in pack

Add measures to package when you add product to pack
```

**Comment by RodrigoBM** [comment]
> @RodrigoBM As a general rule, can you keep commit subject (first line) short and put description in the body (separated by a blank line from the subject).
> 
> Like:
> 
> ```
> delivery_carrier_package_measure_required: put in pack
> 
> Add measures to package when you add product to pack
> ```

Done. I will take this tip in consideration in the next contributions.

**Review by jbaudoux — APPROVED** [review]
Thanks
